### PR TITLE
chore: Merge in contents of `build-nargo` repository

### LINKED ISSUE #2188: Integrate `noir-lang/build-nargo` into this repository
### Problem

The Nargo release process is currently split off into the `noir-lang/build-nargo` repository which handles building Nargo binaries and adding them to release tags. This separation causes issues as we've had a good number of times where we create a release and it's unable to create binaries due to issues which only occur in the release flow.

### Happy Case

We should integrate those workflows into the CI of this repository. We can then use the built binaries to implement #2013 and also run the full cross-compilation flows on any release-please PR which while block any broken release.

### Alternatives Considered

#1921 uses the output of one of these release flows already so we can start amortising these build times across CI workflows.

### Additional Context

_No response_

### Would you like to submit a PR for this Issue?

No

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by TomAFrench**
# Description

## Problem\*

Resolves #2188 

## Documentation

- [ ] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

# PR Checklist\*

- [ ] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: .github/NIGHTLY_BUILD_FAILURE.md
Comment (kevaundray) [inline .github/NIGHTLY_BUILD_FAILURE.md:3]: Can you add me here too -- I'd like to be notified when nightly builds fail too!

**Comment by TomAFrench** [comment]
This PR is currently in draft as I perform testing.

**Comment by TomAFrench** [comment]
We need to clone the noir repo into the `./noir` path as we need to be able to place the wasm binary outside of the repo to keep the repo clean.

**Comment by kevaundray** [comment]
In a separate PR, we can make it such that each night, a new tag is created using the date like in Rust such that we can do the following: `noirup -v 2023-08-08`

**Comment by TomAFrench** [comment]
This should be ready now.

File: .github/workflows/publish.yml
Comment (TomAFrench) [inline .github/workflows/publish.yml:340]: Actually, we can remove this now right @phated?

File: .github/Cross.toml
Comment (kevaundray) [inline .github/Cross.toml:4]: Is this needed?

File: .github/workflows/publish.yml
Comment (kevaundray) [inline .github/workflows/publish.yml:127]: Noting for reference that on Mac we need to use 7z because just using tar for some reason causes the output to be malformed

File: .github/workflows/publish.yml
Comment (kevaundray) [reply .github/workflows/publish.yml:340]: Yes, you can also remove the repo_name since you are running it within noir-lang/noir

File: release-tests/test/6_array.test.js
Comment (kevaundray) [inline release-tests/test/6_array.test.js:5]: Note: We only test 6_array as a smoke test, this was chosen arbitrarily

File: .github/workflows/publish.yml
Comment (TomAFrench) [reply .github/workflows/publish.yml:340]: I don't think that we can remove it entirely due to permissions.

**Review by kevaundray — DISMISSED** [review]
This looks good to me, can change to GITHUB_TOKEN in a separate PR since there are concerns

**Comment by TomAFrench** [comment]
Yeah, I guess we can always trigger a nightly release immediately after merging and seeing if it succeeds.

File: .github/NIGHTLY_BUILD_FAILURE.md
Comment (phated) [inline .github/NIGHTLY_BUILD_FAILURE.md:1]: Is this file still needed if we are checking that binaries can be built in the MQ?

File: .github/workflows/publish.yml
Comment (phated) [inline .github/workflows/publish.yml:17]: Should we instead make this a `merge_queue` only check?

File: .github/workflows/publish.yml
Comment (phated) [inline .github/workflows/publish.yml:69]: Unnecessary if we make these checks required in the MQ, right?

File: .github/workflows/publish.yml
Comment (phated) [inline .github/workflows/publish.yml:85]: Why check it out into a `./noir` directory if we are in the repo now?

File: .github/workflows/publish.yml
Comment (phated) [inline .github/workflows/publish.yml:116]: See above. Might not be needed

File: .github/workflows/publish.yml
Comment (phated) [inline .github/workflows/publish.yml:123]: See above. Might not be needed

File: .github/workflows/publish.yml
Comment (phated) [inline .github/workflows/publish.yml:138]: See above. Might not be needed

File: .github/workflows/publish.yml
Comment (phated) [inline .github/workflows/publish.yml:133]: Need to adjust this if we drop working-dir

File: .github/workflows/publish.yml
Comment (phated) [inline .github/workflows/publish.yml:151]: This can use `GITHUB_SECRET` now, right?

File: .github/workflows/publish.yml
Comment (phated) [inline .github/workflows/publish.yml:152]: Will need to change path if we drop working-dir

File: .github/workflows/publish.yml
Comment (phated) [inline .github/workflows/publish.yml:166]: Won't need if we verify this on MQ

File: .github/workflows/publish.yml
Comment (phated) [inline .github/workflows/publish.yml:213]: See above. Might not be needed

File: .github/workflows/publish.yml
Comment (phated) [inline .github/workflows/publish.yml:189]: See above. Might not be needed

File: .github/workflows/publish.yml
Comment (phated) [inline .github/workflows/publish.yml:221]: See above. Might not be needed

File: .github/workflows/publish.yml
Comment (phated) [inline .github/workflows/publish.yml:231]: Will need to adjust

File: .github/workflows/publish.yml
Comment (phated) [inline .github/workflows/publish.yml:236]: See above. Might not be needed

File: .github/workflows/publish.yml
Comment (phated) [inline .github/workflows/publish.yml:249]: GITHUB_SECRET

File: .github/workflows/publish.yml
Comment (phated) [inline .github/workflows/publish.yml:265]: Won't need with MQ checks

File: .github/workflows/publish.yml
Comment (phated) [inline .github/workflows/publish.yml:281]: See above. Might not be needed

File: .github/workflows/publish.yml
Comment (phated) [inline .github/workflows/publish.yml:305]: See above. Might not be needed

File: .github/workflows/publish.yml
Comment (phated) [inline .github/workflows/publish.yml:250]: Will need to change

File: .github/workflows/publish.yml
Comment (phated) [inline .github/workflows/publish.yml:312]: See above. Might not be needed

File: .github/workflows/publish.yml
Comment (phated) [inline .github/workflows/publish.yml:322]: Will need to change wthout working-dir

File: .github/workflows/publish.yml
Comment (phated) [inline .github/workflows/publish.yml:326]: See above. Might not be needed

File: .github/workflows/publish.yml
Comment (phated) [reply .github/workflows/publish.yml:340]: As commented above, I think GITHUB_TOKEN works here. As with all our workflows, we should also be adding the `permissions` top-level section to the yaml files to know we are requesting the right perms, and only the right perms

File: .github/workflows/publish.yml
Comment (phated) [inline .github/workflows/publish.yml:341]: Will need to change

File: .github/workflows/publish.yml
Comment (phated) [inline .github/workflows/publish.yml:356]: Won't be needed with MQ checks

File: .github/workflows/release.yml
Comment (phated) [inline .github/workflows/release.yml:60]: I don't this this actually works because we don't specify the `workflow_call` event. I also believe that it would need to take a secret passed through because I don't think it inherits secrets, but I'm not clear on that.

I'd probably leave this to a workflow_dispatch even if it is just dispatching to the same repo (it should be able to use `GITHUB_TOKEN` now though

File: release-tests/package.json
Comment (phated) [inline release-tests/package.json:1]: We don't want these in `crates/nargo_cli/tests`?

File: release-tests/yarn.lock
Comment (phated) [inline release-tests/yarn.lock:1]: Nit: Can we drop all dependencies and just use node's built-in test runner and assertion library? Seems like a more streamlined pattern that trying to keep a yarn.lock up to date.

**Review by phated — CHANGES_REQUESTED** [review]
This looks pretty good but I had some feedback on ways it can be cleaned up.

File: .github/workflows/publish.yml
Comment (TomAFrench) [reply .github/workflows/publish.yml:85]: Yeah, I thought the same. The issue is that if we checkout the repo into the runner's workspace root then we don't have anywhere to place the bberg wasm outside of the repo. We can't place it inside of the repository as otherwise we get a dirty build (and tests fail).

Writing this out though we could probably just add a directory to the gitignore and make sure to download any artifacts into there.

File: release-tests/yarn.lock
Comment (TomAFrench) [reply release-tests/yarn.lock:1]: Yeah, I don't like `zx` etc but I think we can remove this in later PRs.

File: .github/workflows/publish.yml
Comment (TomAFrench) [reply .github/workflows/publish.yml:17]: I think we'd want feedback on whether a release is going to fail before we actually try to run the release.

Comment (phated) [reply .github/workflows/publish.yml:17]: That'd be in the merge_queue though, right? The release PR also goes through the MQ

Comment (phated) [reply .github/workflows/publish.yml:17]: Such that we know every PR that lands into master will allow a release to be built, including the release PR itself.

Comment (TomAFrench) [reply .github/workflows/publish.yml:17]: My point is I think we'd want that feedback before it gets to the MQ so we _know_ that the MQ will pass.

Comment (phated) [reply .github/workflows/publish.yml:17]: I generally agree, but I think @kevaundray would want this check to only happen at the MQ since it takes longer than other CI stuff.

If he also wants it to run against PRs, we'd want to use `pull_request` **and** `merge_queue` events like we use on other CI stuff now.

Comment (TomAFrench) [reply .github/workflows/publish.yml:17]: Ah, I get what you mean. :facepalm:

Comment (TomAFrench) [reply .github/workflows/publish.yml:17]: Yeah, that seems sensible.

File: .github/workflows/publish.yml
Comment (TomAFrench) [reply .github/workflows/publish.yml:69]: Yeah, the stars would really need to align for this to be triggered.

File: .github/workflows/release.yml
Comment (TomAFrench) [reply .github/workflows/release.yml:60]: Ah, `workflow_call != workflow_dispatch`. Good catch.

File: .github/workflows/publish.yml
Comment (kobyhallx) [inline .github/workflows/publish.yml:90]: Should we bump to 14.3.1?
https://github.com/actions/runner-images/blob/main/images/macos/macos-13-Readme.md | grep 14E300c

File: release-tests/package.json
Comment (kobyhallx) [reply release-tests/package.json:1]: I think we do, we would also want to extend over couple other interesting cases.

File: release-tests/yarn.lock
Comment (kobyhallx) [reply release-tests/yarn.lock:1]: zx is there to help you with bunch of os stuff you would otherwise code by hand...

File: release-tests/package.json
Comment (TomAFrench) [reply release-tests/package.json:1]: The integration tests now use the CLI to call the nargo binary so the `6_array.test.js` is kinda redundant. The only meaningful test we're doing here is checking that
- Some version gets inserted into the binary (not validated to be correct)
- The binary was built from a clean commit.

File: .github/workflows/publish.yml
Comment (TomAFrench) [reply .github/workflows/publish.yml]: Not sure on the significance of this but bumped.

File: .github/workflows/publish.yml
Comment (phated) [inline .github/workflows/publish.yml:41]: We should not use an authToken here because this will be run on merge_group

**Review by phated — CHANGES_REQUESTED** [review]
Looking good. Just a couple more things and a clarification.

File: release-tests/package.json
Comment (TomAFrench) [reply release-tests/package.json:1]: My point was that more that we're likely going to be reworking these tests shortly after this PR gets merged anyway due to changes in the nargo integration tests since they were made. We can then move them, remove `zx`, etc. as we like but I'd prefer to focus this PR on just replicating `build-nargo`'s functionality.

I don't think that `crates/nargo_cli/tests` makes sense for these as they're distinct from the integration tests. The integration tests test the functionality of the codebase whereas these perform checks on a release artifact and so I would expect these to fail under conditions which integration tests should not fail.

File: .github/Cross.toml
Comment (TomAFrench) [reply .github/Cross.toml:4]: :shrug: I'm just aiming to maintain the same behaviour as current in this PR. I think we can perform tweaks in followups.

Comment (TomAFrench) [reply .github/Cross.toml:4]: https://github.com/noir-lang/noir/issues/2245

File: .github/workflows/publish.yml
Comment (TomAFrench) [inline .github/workflows/publish.yml:18]: Needs removal.

File: .github/workflows/publish.yml
Comment (phated) [inline .github/workflows/publish.yml:22]: A future improvement we can make to this is to apply it only to a job and moving the release upload into its own job that relies on the various build jobs

File: .github/workflows/publish.yml
Comment (phated) [inline .github/workflows/publish.yml:26]: If we make a nix dev env that contains cargo-cross and bb wasm, we wouldn't need this. Just a thought about future optimization.

File: .github/workflows/publish.yml
Comment (phated) [inline .github/workflows/publish.yml:129]: Can we create an issue to remove all deps (using just node core functionality) to avoid yarn install and a yarn.lock?

File: .github/workflows/release.yml
Comment (phated) [inline .github/workflows/release.yml:58]:
```suggestion
```
I don't think we need this todo

File: .github/workflows/release.yml
Comment (phated) [inline .github/workflows/release.yml:66]: I'm not sure everywhere this token might be used, but if this was the only place, we'd want to purge it from the system

File: release-tests/package.json
Comment (phated) [inline release-tests/package.json:5]:
```suggestion
```
Might as well remove this field, yeah?

File: release-tests/test/6_array.test.js
Comment (phated) [inline release-tests/test/6_array.test.js:15]: Future improvement is that you can actually just pass a flag to node that does this

**Review by phated — DISMISSED** [review]
Just some thoughts on future improvements. It'd be nice to have all these as issues 🙏

File: .github/workflows/release.yml
Comment (TomAFrench) [reply .github/workflows/release.yml:66]: `NOIR_REPO_TOKEN` you mean?

File: release-tests/package.json
Comment (TomAFrench) [reply release-tests/package.json:5]: It avoid a warning and there's no harm.

**Comment by TomAFrench** [comment]
Issue for comments related to `release-tests` opened here https://github.com/noir-lang/noir/issues/2262

File: release-tests/package.json
Comment (phated) [reply release-tests/package.json:5]: If you want to keep it, I think it should match noir's license and be the SPDX for "Apache or MIT" so as not confuse anyone looking around the project.

File: .github/workflows/release.yml
Comment (phated) [reply .github/workflows/release.yml:66]: Yeah, the more we are able to use `GITHUB_TOKEN`, the better. So we just need to remove any unused custom secrets. Probably need to grep other workflows for`NOIR_REPO_TOKEN`

Comment (TomAFrench) [reply .github/workflows/release.yml:66]: Yeah, we only use it for triggering the `noir_wasm` release workflow but that should be handled with https://github.com/noir-lang/noir/issues/2210

Comment (TomAFrench) [reply .github/workflows/release.yml:66]: Actually.... that shouldn't need this token at all.

**Review by phated — APPROVED** [review]

**Pull request merged**
